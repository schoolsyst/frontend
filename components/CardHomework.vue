<template>
  <base-card :class="{ '--card-homework': true, expanded }">
    <div class="header">
      <badge-subject
        :color="homework.subject.color"
        editable
        expandable
        @edit="editSubject"
        >{{ homework.subject.name }}</badge-subject
      >
      <p class="title" @input="$emit('input', { title: $event })">
        {{ homework.title }}
      </p>
      <button-icon class="toggle-expand" @click="toggleExpand"
        >keyboard_arrow_down</button-icon
      >
    </div>
    <div class="details">
      <p>{{ homework.details }}</p>
    </div>
    <transition name="slide">
      <div key="1" class="actions" v-if="!editing">
        <button-icon @click="deleteHomework">delete</button-icon>
        <button-icon @click="editHomework">edit</button-icon>
        <text-link
          v-if="homework.quizzes.length > 0"
          :to="`/learn/${homework.uuid}`"
          >RÃ©viser</text-link
        >
      </div>
      <div key="2" class="actions" v-else>
        <button-primary @click="finishEditing" small>OK</button-primary>
      </div>
    </transition>
  </base-card>
</template>

<script lang="ts">
// eslint-disable-next-line no-unused-vars
import Vue, { PropOptions } from 'vue'
import BaseCard from '~/components/BaseCard.vue'
import BadgeSubject from '~/components/BadgeSubject.vue'
import ButtonPrimary from '~/components/ButtonPrimary.vue'
import ButtonIcon from '~/components/ButtonIcon.vue'
import TextLink from '~/components/TextLink.vue'
// import '~/types/api-resources'

export default Vue.extend({
  components: { BaseCard, BadgeSubject, ButtonPrimary, ButtonIcon, TextLink },
  props: {
    homework: {
      type: Object,
      required: true,
    } as PropOptions<ApiResourceHomework>,
  },
  data() {
    return {
      editing: false,
      saving: false,
      deleting: false,
      expanded: false,
    }
  },
  methods: {
    editHomework(): void {
      this.editing = true
    },
    async finishEditing(): Promise<void> {
      this.saving = true
      await this.$store.dispatch('homework/edit', { modifications: {} })
      // await this.$store.dispatch('homework/load')
      this.saving = false
      this.editing = false
    },
    editSubject(): void {
      // this.$emit('input', {
      //   subject: this.$pickSubject({ selected: this.homework.uuid }),
      // })
    },
    async deleteHomework(): Promise<void> {
      this.deleting = true
      await this.$store.dispatch('homework/delete', {
        uuid: this.homework.uuid,
      })
      this.deleting = false
    },
    async toggleExpand(): Promise<void> {
      this.expanded = !this.expanded
      if (!this.expanded && this.editing) {
        await this.finishEditing()
      }
    },
  },
})
</script>

<style lang="stylus" scoped>
//
//Definitions
//

//
//Positioning
//
.header
  display flex
  align-items center

.toggle-expand
  margin-left auto

.actions
  display flex
  justify-content end
  align-items center

//
//Sizing
//

.details
  font-size 0.75em

//
//Spacing
//
.title
  margin-left 0.5em

.details
  margin-top 0.5em

.actions
  margin-top 0.5em

.actions > *:not(:last-child)
  margin-right 0.7em

//
//Decoration
//

//
//Colors
//

//
//Typography
//

//
//States
//

//// Expanded/not expanded
.--card-homework, .actions, .details, .toggle-expand
  transition ease 250ms all

.--card-homework:not(.expanded)
  max-height 5.1rem

  .actions
    opacity 0

  .details
    line-clamp 1

  .toggle-expand
    transform rotate(0)

.--card-homework.expanded
  max-height 100rem

  .actions
    opacity 1

  .details
    line-clamp 0


  .toggle-expand
    transform rotate(180deg)


.slide-enter-active, .slide-leave-active
    transition all 250ms ease
.slide-enter, .slide-eave-to
    opacity: 0 !important
</style>
