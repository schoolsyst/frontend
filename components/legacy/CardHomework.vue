<template lang="pug">
    .card(
        @click="$emit('click')"
        :class="{completed, opened, [`is-a-${type.toLowerCase()}`]: true}"
        v-touch:swipe.left="toggleComplete"
        v-touch:swipe.right="toggleComplete"
    )
        .complete-slider(@click.stop="toggleComplete" :title="sliderTooltip")
            Icon(v-show="completed") refresh
            Icon(v-show="!completed") check
        .infos
            .first-line
                .strikethrough-line
                BadgeSubject.subject-color(v-bind="subject" variant="dot")
                span.name {{ name }}
                Icon.graded-indicator(
                  v-if="['COURSEWORK', 'TEST'].includes(type)"
                  v-tooltip="type === 'TEST' ? 'Contrôle' : 'Noté'"
                ) error_outline
            .details(v-if="details" v-html="details")
</template>

<script>
import { mapActions } from 'vuex'
import debounce from 'lodash.debounce'
import Icon from '~/components/legacy/Icon.vue'
import BadgeSubject from '~/components/legacy/BadgeSubject.vue'
import InputField from '~/components/legacy/InputField.vue'
export default {
  components: { Icon, BadgeSubject, InputField },
  props: {
    name: {
      type: String,
      required: true,
    },
    details: {
      type: String,
      default: '',
    },
    subject: {
      type: Object,
      required: true,
    },
    uuid: {
      type: String,
      required: true,
    },
    progress: {
      type: Number,
      default: 0,
    },
    opened: {
      type: Boolean,
      default: false,
    },
    type: {
      type: String,
      required: true,
    },
  },
  computed: {
    completed() {
      return this.progress >= 1
    },
    sliderTooltip() {
      return `Marquer comme ${this.completed ? 'non-' : ''}terminé`
    },
  },
  methods: {
    ...mapActions('homework', ['switchCompletion']),
    toggleComplete: debounce(
      async function() {
        this.$emit('completion-switch', !this.completed)
        await this.switchCompletion({ uuid: this.uuid })
      },
      500,
      { leading: true, trailing: false }
    ),
  },
}
</script>

<style lang="stylus" scoped>
.card
  display flex
  overflow hidden
  max-width 100%
  //height: 100px
  width 500px
  border-radius var(--border-radius)
  cursor pointer
  transition box-shadow 0.25s ease

.strikethrough-line
  position absolute
  top 50%
  left calc(((100% - (100% - (var(--strikethrough-line-offset) * 2))) / 2))
  width 0
  height 0.1em
  background var(--black)
  transition width 0.25s ease
  --strikethrough-line-offset 0

.card.completed .complete-slider
  background-color var(--blue)

.card:not(.completed) .complete-slider
  background-color var(--green)

.complete-slider
  z-index 10
  display flex
  flex-shrink 0
  align-items center
  overflow hidden
  width 0
  border-radius var(--border-radius)
  border-top-right-radius 0
  border-bottom-right-radius 0
  color var(--white)
  transition width 0.25s ease

  i
    margin-left 0.5em

.infos
  position relative
  display flex
  flex-direction column
  padding 1.1em 1.2em
  width 100%
  border 1px solid var(--grey-light)
  border-radius var(--border-radius)

  .first-line
    position relative
    display flex
    align-items center
    overflow hidden
    width 100%
    text-overflow ellipsis
    white-space nowrap

  .name
    margin-left 0.5rem

  .details
    flex-grow 0
    overflow hidden
    margin-top 0.5em
    width 100%
    text-overflow ellipsis
    white-space pre-wrap
    font-size 0.75em

  .graded-indicator
    margin-left auto
    font-size 1.3em

.is-a-test .graded-indicator
  color var(--red)

.card.completed
  .infos
    opacity 0.5

  .strikethrough-line
    width calc(100% - (var(--strikethrough-line-offset) * 2))

.card:hover
  &
    box-shadow var(--shadow-2)

  .complete-slider
    width 3em

  &.completed .complete-slider:hover
    background-color var(--blue-dark)

  &:not(.completed) .complete-slider:hover
    background-color var(--green-light)

  .infos
    border-left 0
    border-top-left-radius 0
    border-bottom-left-radius 0
</style>
