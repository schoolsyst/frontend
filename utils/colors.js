import tinycolor from 'tinycolor2'

export const lightColor = baseColor => {
  /* see https://codepen.io/ewen-lbh/pen/dyYKWVL */
  const { h, s } = tinycolor(baseColor).toHsl()
  return tinycolor({ h, s: Math.min(2 * s, 1), l: 0.85 }).toHslString()
}
export const darkColor = baseColor => {
  /* see https://codepen.io/ewen-lbh/pen/dyYKWVL */
  const { h, s } = tinycolor(baseColor).toHsl()
  return tinycolor({ h, s: Math.min(0.25 * s, 1), l: 0.3 }).toHslString()
}

export const blackOrWhite = bgColor =>
  tinycolor(bgColor).isLight() ? 'black' : 'white'
