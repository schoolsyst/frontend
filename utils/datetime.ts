export const removeByProp = (
  arrayOfObjects: Record<string, any>[],
  propName: string,
  propValue: any
) => {
  const arr = [...arrayOfObjects]
  return arr.splice(
    arr.findIndex(obj => obj[propName] === propValue),
    1
  )
}
