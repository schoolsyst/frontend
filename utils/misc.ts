// Reset time component to 00:00:00
// Useful to do date-wise comparison, w/o taking time into account.
export const removeTime = (date: Date) => {
  date.setHours(0)
  date.setMinutes(0)
  date.setSeconds(0)
  return date
}
