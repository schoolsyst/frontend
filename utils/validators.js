export const isHexColor = value => /^#([0-9a-fA-F]{3}){1,2}$/.test(value)
export const isURL = value =>
  /^(ftp|https?):\/\/([\w\d-]+\.)+\w{2,}(\/.+)?$/.test(value)
export const isIn = allowedValues => value => allowedValues.includes(value)
export const isUUID = value =>
  /[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/.test(value)
