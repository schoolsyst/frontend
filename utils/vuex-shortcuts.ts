import { isBefore } from 'date-fns'
type Gender = 'F' | 'M'
type CheckerName =
  | 'maximum'
  | 'minimum'
  | 'maxLength'
  | 'required'
  | 'notEmpty'
  | 'isAWeekType'
  | 'before'
  | 'isAnEmail'
  | 'isOfType'
  | 'isAColor'

export const getValidator = ({
  constraints,
  resourceName,
  fieldNames,
  customConstraints = [],
  debug,
}: {
  constraints: Record<CheckerName, Record<any, string[]> | string[]>
  resourceName: {
    name: string
    gender: Gender
    article: string | null
    indeterminateArticle: string | null
  }
  fieldNames: Record<string, { name: string; gender: Gender }>
  customConstraints: {
    message: string
    field: string | null
    constraint: (getters: object, object: object) => boolean
  }[]
  debug: boolean
}) => getters => (object: Record<string, any>) => {
  debug = debug || false
  if (process.env.NODE_ENV !== 'development') debug = false
  /* Factory to create a validator.
  Describe constraints on fields, error messages are generated automatically.
  */

  if (debug) {
    console.group(`[validator] validating resource "${resourceName.name}"`) // eslint-disable-line
    console.log(`Constraints:`) // eslint-disable-line
    console.log({ ...constraints, ...customConstraints }) // eslint-disable-line
    console.log(`Fields:`) // eslint-disable-line
    console.log(
      // eslint-disable-line
      Object.fromEntries(
        Object.keys(fieldNames).map(field => [field, object[field]])
      )
    )
  }

  // Article in french
  const article = (noun: string, feminine: boolean, indeterminate = false) => {
    const vowels = ['a', 'e', 'i', 'o', 'u', 'y']
    if (vowels.includes(noun[0]) && !indeterminate) return "l'"
    else if (feminine) return indeterminate ? 'une ' : 'la '
    else return indeterminate ? 'un ' : 'le '
  }

  // Resource name article & indeterminate article
  resourceName.article = article(
    resourceName.name,
    resourceName.gender === 'F',
    false
  )
  resourceName.indeterminateArticle = article(
    resourceName.name,
    resourceName.gender === 'F',
    true
  )

  // Uppercase first char
  const upperFirst = (string: string) => string[0].toUpperCase() + string.slice(1)

  // Checkers

  const checkers: {
    [key in CheckerName]: ({
      arg,
      fieldName,
    }: {
      arg: any
      fieldName: string
    }) => boolean
  } = {
    maximum: ({ arg, fieldName }) =>
      object[fieldName] === null ||
      (typeof object[fieldName] === 'number' && object[fieldName] <= arg),
    minimum: ({ arg, fieldName }) =>
      object[fieldName] === null ||
      (typeof object[fieldName] === 'number' && object[fieldName] >= arg),
    maxLength: ({ arg, fieldName }) =>
      typeof object[fieldName] === 'string'
        ? object[fieldName].length <= arg
        : true,
    required: ({ arg, fieldName }) =>
      Object.prototype.hasOwnProperty.call(object, fieldName) &&
      object[fieldName] !== null,
    notEmpty: ({ arg, fieldName }) =>
      object[fieldName] && object[fieldName].trim().length > 0,
    isAWeekType: ({ arg, fieldName }) =>
      ['Q1', 'Q2', 'BOTH'].includes(object[fieldName]),
    before: ({ arg, fieldName }) => isBefore(object[fieldName], object[arg]),
    isAColor: ({ arg, fieldName }) =>
      object[fieldName].match(/^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$/) !== null,
    isAnEmail: ({ arg, fieldName }) =>
      // regex source: https://emailregex.com/
      typeof object[fieldName] === 'string' &&
      object[fieldName].match(
        /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
      ) !== null,
    isOfType: ({ arg, fieldName }) =>
      // eslint-disable-next-line valid-typeof
      typeof object[fieldName] === arg,
  }
  const check = ({
    errorName,
    fieldName,
    errorArg,
  }: {
    errorName: CheckerName
    fieldName: string
    errorArg: any
  }) => {
    // Wrap checkers with object property existential check
    /* Special case for required, which checks if the property exist
       and returnes *false* instead.
    */
    if (
      errorName !== 'required' &&
      !Object.prototype.hasOwnProperty.call(object, fieldName)
    ) {
      return true
    }
    return checkers[errorName]({ arg: errorArg, fieldName })
  }

  // Error messages
  const message = ({
    errorName,
    errorArg,
    fieldName,
  }: {
    errorName: CheckerName
    errorArg: any
    fieldName: string
  }): string => {
    const { gender, name } = fieldNames[fieldName]
    const fieldIsFeminine = gender === 'F'
    const determinateArticle = article(name, fieldIsFeminine, false)
    const indeterminateArticle = article(name, fieldIsFeminine, true)
    const argNameWithArticle =
      article(
        fieldNames[fieldName].name,
        fieldNames[fieldName].gender === 'F',
        false
      ) + fieldNames[fieldName].name
    const fieldNameWithArticle = determinateArticle + name
    return upperFirst(
      {
        maximum: `${fieldNameWithArticle} ne doit pas dépasser ${errorArg}`,
        minimum: `${fieldNameWithArticle} doit être d'au moins ${errorArg}`,
        maxLength: `${fieldNameWithArticle} ne doit pas dépasser ${errorArg} caractère${
          errorArg === 1 ? '' : 's'
          }`,
        required: `Veuillez renseigner ${indeterminateArticle}${name}`,
        isAWeekType: `${fieldNameWithArticle} doit être paire, impaire ou les deux.`,
        before: `${fieldNameWithArticle} doit être avant ${argNameWithArticle}`,
        isAColor: `${fieldNameWithArticle} doit être une couleur au format hexadécimal. Exemple: #09f ou #0479cf`,
        notEmpty: `${fieldNameWithArticle} est requis${
          fieldIsFeminine ? 'e' : ''
          }`,
        isAnEmail:
          name === 'adresse email'
            ? `${fieldNameWithArticle} doit être valide`
            : `${fieldNameWithArticle} doit être une adresse email valide`,
      }[errorName]
    )
  }

  // Stores the errors
  const errorMessages: Record<string, string[]> = {}
  // Fill each field with an empty array
  Object.keys(fieldNames).forEach(field => {
    errorMessages[field] = []
  })
  let validated = true

  // Go through the constraints
  for (const errorName in constraints) {
    if (Object.prototype.hasOwnProperty.call(constraints, errorName)) {
      const fieldsOrArgs = constraints[errorName]
      // No arguments, field names are passed directly
      // eslint-disable-next-line
      if (fieldsOrArgs instanceof Array) {
        fieldsOrArgs.forEach(fieldName => {
          if (!check({ errorName, fieldName, errorArg: null })) {
            // Some error occured, add the error message
            errorMessages[fieldName].push(
              message({ errorName, fieldName, errorArg: null })
            )
            validated = false
          }
        })
      }
      // Error arguments, multiple cases. eg: maximum
      else {
        for (const errorArg in fieldsOrArgs) {
          if (Object.prototype.hasOwnProperty.call(fieldsOrArgs, errorArg)) {
            const fields = fieldsOrArgs[errorArg]
            fields.forEach(fieldName => {
              if (!check({ errorName, fieldName, errorArg })) {
                errorMessages[fieldName].push(
                  message({ errorName, fieldName, errorArg })
                )
                validated = false
              }
            })
          }
        }
      }
    }
  }

  // Go through custom constraints
  customConstraints.forEach(({ message, constraint, field }) => {
    if (!constraint(getters, object)) {
      field = field === null ? 'nonFieldErrors' : field
      if (Object.prototype.hasOwnProperty.call(errorMessages, field)) {
        errorMessages[field].push(message)
      } else {
        errorMessages[field] = [message]
      }
      validated = false
    }
  })

  // Creates a single string that reports all of the errors.
  // Get the total errors count
  let flatErrorMessage = ''
  let errorsCount = 0
  for (const field in errorMessages) {
    if (Object.prototype.hasOwnProperty.call(errorMessages, field)) {
      const errors = errorMessages[field]
      if (errors.length) {
        flatErrorMessage += errors.join(', ') + '\n'
        errorsCount += errors.length
      }
    }
  }

  // returns the object
  const ret = {
    validated,
    errors: errorMessages,
    message: flatErrorMessage,
    count: errorsCount,
  }
  if (debug) {
    if (ret.validated) {
      console.log('Validated.')
    } else {
      console.log('Errors:')
      Object.entries(ret.errors).forEach(([name, errors]) => {
        if (errors.length) {
          console.log(`    ${name}:`)
          console.log(errors)
        }
      })
    }
    console.groupEnd()
  }

  return ret
}

export const getMutations = (
  what,
  mapWith = o => o,
  pure = true,
  verbs = ['add', 'set', 'del', 'patch'],
  primaryKey = 'uuid',
  debug = false
) => {
  /* Factory to create basic mutations while preserving DRYness of code
   * set, add, del & patch are booleans that—when set to false—disable
   * a particular mutation from being built & returned.
   *
   * @param pure: makes the function return solely the verb as the function name.
   *              useful for store modules consisting of *one* resource only.
   *              example: {SET:..., ADD:... ...} instead of {SET_THINGS:..., ADD_THING:... ...}
   *
   * @param what: the name of the resource. Will be used to calculate the
   *              state's array & mutations' names.
   * @param mapWith: a function to map each object to when putting into the state
   *                 Defaults to `o => o`
   * @param primaryKey: the name of the primary key used for mutations acting on a single object.
   * @param debug: Log every mutation if true
   */

  const mutations = {}
  const WHAT = pure ? '' : '_' + constantCase(what)
  const whats = what + 's'

  if (verbs.includes('set'))
    mutations[`SET${WHAT}${pure ? '' : 'S'}`] = (state, items) => {
      if (debug) {
        console.group(`${whats}/SET`)
        console.log('before:')
        console.log(state[whats])
      }
      Vue.set(state, whats, items.map(mapWith))
      if (debug) {
        console.log('after:')
        console.log(state[whats])
        console.groupEnd()
      }
    }
  if (verbs.includes('add'))
    mutations[`ADD${WHAT}`] = (state, item) => {
      if (debug) {
        console.group(`${whats}/ADD`)
        console.log('item:')
        console.log(item)
        console.log('before:')
        console.log(state[whats])
      }
      Vue.set(state, whats, [...state[whats], mapWith(item)])
      if (debug) {
        console.log('after:')
        console.log(state[whats])
        console.groupEnd()
      }
    }
  if (verbs.includes('del'))
    mutations[`DEL${WHAT}`] = (state, pk) => {
      if (debug) {
        console.group(`${whats}/DEL`)
        console.log('item:')
        console.log(state[whats][pk])
        console.log('before:')
        console.log(state[whats])
      }
      Vue.set(
        state,
        whats,
        state[whats].filter(o => o[primaryKey] !== pk)
      )
      if (debug) {
        console.log('after:')
        console.log(state[whats])
        console.groupEnd()
      }
    }
  if (verbs.includes('patch'))
    mutations[`PATCH${WHAT}`] = (state, { pk, modifications }) => {
      if (debug) {
        console.group(`${whats}/PATCH`)
        console.log('before:')
        console.log(state[whats])
      }
      // Get the requested item's index in the state array
      const idx = state[whats].map(o => o[primaryKey]).indexOf(pk)
      // Apply modifications
      let item = { ...state[whats][idx], ...modifications }
      // Re-run mapWith
      item = mapWith(item)
      // Set in store
      Vue.set(state[whats], idx, item)
      if (debug) {
        console.log('item:')
        console.log(item)
        console.log('after:')
        console.log(state[whats])
        console.groupEnd()
      }
    }

  return mutations
}
