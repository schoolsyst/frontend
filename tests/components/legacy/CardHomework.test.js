import { bootstrapComponentTest } from '../../utils'
import CardHomework from '~/components/legacy/CardHomework.vue'
// import homework from '~/store/homework'

const mockHomework = {
  uuid: 'c0ffee-daedead-b16b00b5',
  subject: {
    name: 'Subject',
    color: '#c0ffee',
  },
  progress: 0.3,
  type: 'HOMEWORK',
  name: 'Lorem ipsum',
  details:
    'Lorem ipsum dolor sit amet consecitur nae doe flat is justice et deald.',
}
const homeworkStore = {
  actions: {
    switchCompletion: jest.fn(),
  },
}

const { mnt, store } = bootstrapComponentTest(
  CardHomework,
  { homework: homeworkStore },
  mockHomework
)

describe('<CardHomework>', () => {
  it('is a Vue instance', () => {
    const CardHomework = mnt()
    expect(CardHomework.isVueInstance()).toBeTruthy()
  })

  describe('completed state', () => {
    const CardHomework = mnt({ progress: 1 })
    it('matches snapshot', () => {
      expect(CardHomework).toMatchSnapshot()
    })
  })

  describe('uncompleted state', () => {
    const CardHomework = mnt({ progress: 0 })
    it('matches snapshot', () => {
      expect(CardHomework).toMatchSnapshot()
    })
  })

  it('calls switchCompletion when the complete slider is clicked', () => {
    const CardHomework = mnt({ progress: 0 })
    CardHomework.get('.complete-slider').trigger('click')
    setTimeout(() => {
      expect(CardHomework.emitted('completion-switch')).toBeTruthy()
      expect(CardHomework.emitted('completion-switch')[1]).toEqual(true)
      expect(store.dispatch).toHaveBeenCalledWith('homework/switchCompletion', {
        uuid: mockHomework.uuid,
      })
    }, 500)
  })

  it("doesn't call switchCompletion when the card is clicked", () => {
    const CardHomework = mnt()
    CardHomework.trigger('click')
    setTimeout(() => {
      expect(store.dispatch).not.toHaveBeenCalled()
    }, 500)
  })

  it('emits a click event when the card is clicked', () => {
    const CardHomework = mnt()
    CardHomework.trigger('click')
    expect(CardHomework.emitted('click')).toBeTruthy()
  })
})
