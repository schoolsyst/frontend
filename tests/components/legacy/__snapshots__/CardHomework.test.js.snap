// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardHomework> completed state matches snapshot 1`] = `
<div class="card completed is-a-homework">
  <div title="Marquer comme non-terminé" class="complete-slider">
    <icon-stub>refresh</icon-stub>
    <icon-stub style="display: none;">check</icon-stub>
  </div>
  <div class="infos">
    <div class="first-line">
      <div class="strikethrough-line"></div>
      <badgesubject-stub color="#c0ffee" name="Subject" placeholdername="Choisir..." variant="dot" class="subject-color"></badgesubject-stub><span class="name">Lorem ipsum</span>
      <!---->
    </div>
    <div class="details">Lorem ipsum dolor sit amet consecitur nae doe flat is justice et deald.</div>
  </div>
</div>
`;

exports[`<CardHomework> uncompleted state matches snapshot 1`] = `
<div class="card is-a-homework">
  <div title="Marquer comme terminé" class="complete-slider">
    <icon-stub style="display: none;">refresh</icon-stub>
    <icon-stub>check</icon-stub>
  </div>
  <div class="infos">
    <div class="first-line">
      <div class="strikethrough-line"></div>
      <badgesubject-stub color="#c0ffee" name="Subject" placeholdername="Choisir..." variant="dot" class="subject-color"></badgesubject-stub><span class="name">Lorem ipsum</span>
      <!---->
    </div>
    <div class="details">Lorem ipsum dolor sit amet consecitur nae doe flat is justice et deald.</div>
  </div>
</div>
`;
